internal/consensus/state.go

```diff
@@ -3053,7 +3053,10 @@ func (cs *State) signVote(
 	ctxto, cancel := context.WithTimeout(ctx, timeout)
 	defer cancel()
 
-	stateID := cs.StateID()
+	stateID := vote.StateID()
+	if len(hash) == 0 {
+		v.AppHash = nil
```

Comment:
basically, it is not necessary to reset `AppHash`, because `AppHash` is used to sign a state, a combination of height and app-hash, so if hash is nil, then state signing and verification will be skipped